README.md
xkcd-hash
=========

XKCD Hash SKein1024 brute forcing.

Overview
--------

Goal is to find input strings such that skein1024(input)
is as 'close' as possible to the given value.

This is for XKCD's April 1st comic: http://xkcd.com/1193/

Primary UIUC implementation.

Usage
-----

    $ make
    $ ./hash <num_threads>


If unspecified, num_threads is set to the number of available processors.


Compatibility
-------------

This should work on most 32/64bit machines, and has been reported
to work well on both Linux and Mac machines.

On Linux, we use assembly versions of the most compute-intensive part of the
skein hash calculation, but are unable to use this on Mac OS X due to
incompatibility with the gas style assembly.

Porting this to windows should be straightforward, although
things like the way we seed the PRNG will need to be changed.

Fixed Skein Assembly for Linux x32/64
-------------------------------------

The Skein implementation used in this project is from the most recent
NIST submission, making use of the assembly implementation when possible.
As a byproduct of this competition, I ended up fixing a few minor bugs that caused
their implementation to generate incorrect hashes or crash.  These changes
may be use beyond this project.

String Generation Details
-------------------------

Our string generation is accomplished by periodically filling
the majority of the string with random characters from our
character set, and exhaustively trying all possibilities
for the remaining characters.

Generating random strings is expensive, requiring many
calls to random().  The exhaustive search is done to amortize
this cost, while still mostly generating random strings.

The repeated random generation is an attempt to prevent
two identical starting states from causing hosts to compute
purely redundant calculations.  This way, happening to output
the same 128 integers is not taken to be a reflection
of having the same internal PRNG state.

Strings of length 128 are used to match the 1024bit block size,
which means we can hash them faster.

For fun, we use a fixed vanity prefix and suffix string in
the generated input strings.  This also aids in demonstrating who was
responsible for generating a particular string.



libbsd-0.4.2/ChangeLog
commit 918a4dba4a35fdc241381a2ff9f1e697653adf88
Author: Guillem Jover <guillem@hadrons.org>
Date:   Wed Jun 27 09:15:15 2012 +0200

    Release libbsd 0.4.2

 src/Makefile.am |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

commit 444bd1fbb8f53047b1ee2b69dc4a8eb65537bc68
Author: Guillem Jover <guillem@hadrons.org>
Date:   Thu Jun 21 10:16:14 2012 +0200

    man: Use minus signs and hyphens consistently
    
    Any string that can be copy & pasted into a terminal, for example,
    needs to correctly use minus signs (escaped dashes), instead of the
    default hyphen.

 man/flopen.3  |    2 +-
 man/pidfile.3 |    4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

commit 1d69ae1cd56855b0e9cf5c40f24854f77d9009ca
Author: Guillem Jover <guillem@hadrons.org>
Date:   Mon Jun 4 05:47:26 2012 +0200

    man: Recode flopen(3) to UTF-8
    
    Spotted-by: Eric Smith <eric@brouhaha.com>

 man/flopen.3 |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

commit 14524b545d7b3f3f8c01405e9ef5ba45d438424c
Author: Guillem Jover <guillem@hadrons.org>
Date:   Sun Jun 3 07:39:42 2012 +0200

    build: Set default compiler variables from configure
    
    This centralizes the setting so there's no duplication anymore,
    makes sure the user supplied variables are never overridden, and
    are only set when using gcc.
    
    Reported-by: Samuli Suominen <ssuominen@gentoo.org>

 configure.ac     |    8 ++++++++
 src/Makefile.am  |    3 ---
 test/Makefile.am |    3 ---
 3 files changed, 8 insertions(+), 6 deletions(-)

commit c21d788fea18ddb8f6eb309c73783df5ccd6bf53
Author: Guillem Jover <guillem@hadrons.org>
Date:   Fri Jun 1 08:20:37 2012 +0200

    Release libbsd 0.4.1

 src/Makefile.am |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

commit fdcae5770799b232d50bf311d69c375f730a1ed3
Author: Guillem Jover <guillem@hadrons.org>
Date:   Fri Jun 1 08:17:42 2012 +0200

    build: Set runtimelibdir to libdir
    
    This makes sure the install-exec-hook under src works as expected even
    when no runtimelibdir was specified, otherwise the symlinks end up
    pointing to non-existing targets.
    
    Reported-by: Ryan Mullen <rmmullen@gmail.com>

 src/Makefile.am |    2 ++
 1 file changed, 2 insertions(+)

commit e9e4a60d7e18c103be6df33cbbeaf249578a1ac5
Author: Guillem Jover <guillem@hadrons.org>
Date:   Fri Jun 1 08:15:00 2012 +0200

    build: Use MKDIR_P variable instead of literal «mkdir -p»

 src/Makefile.am |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

commit 309c82a016d5cfc84d71066f21edc7cd7f5ab95d
Author: Guillem Jover <guillem@hadrons.org>
Date:   Tue May 29 04:38:07 2012 +0200

    Release libbsd 0.4.0

 src/Makefile.am |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

commit cd67cb14173b4329d4cb8c74ea88240e5e4bf2c3
Author: Guillem Jover <guillem@hadrons.org>
Date:   Tue May 29 07:17:46 2012 +0200

    Use implicit <md5.h> from overlay instead explicit <bsd/md5.h>
    
    This was assuming an installed <bsd/md5.h> on the system, due to the
    build system not including -Iinclude/ anymore.
    
    Regression introduced in commit 901ed630fc64fca828e031bc8fa6780db9f05db1.

 src/Makefile.am |    2 +-
 src/hash/md5.c  |    2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

commit 1b5b1cd52ab0209d6213f02ba62fbb959d534fd2
Author: Guillem Jover <guillem@hadrons.org>
Date:   Tue Apr 24 04:20:58 2012 +0200

    Move mdX.3 man page to section 3bsd
    
    This makes sure there will be no collisions on systems where a mdX.3
    provided by a third party is already present.

 man/Makefile.am         |    4 ++--
 man/{mdX.3 => mdX.3bsd} |    2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

commit ddefaae330f95d89cd1046bdde43d51fd0783c1b
Author: Guillem Jover <guillem@hadrons.org>
Date:   Tue Apr 24 04:05:48 2012 +0200

    Do not quote man page titles

 man/flopen.3   |    2 +-
 man/pidfile.3  |    2 +-
 man/strtonum.3 |    2 +-
 man/tree.3     |    2 +-
 4 files changed, 4 insertions(+), 4 deletions(-)

commit e7f3976088b038687e54a2df09cb9e3332af42df
Author: Guillem Jover <guillem@hadrons.org>
Date:   Fri Mar 23 10:31:42 2012 +0100

    Add email address to my name

 COPYING                  |    2 +-
 include/bsd/bsd.h        |    2 +-
 include/bsd/err.h        |    2 +-
 include/bsd/getopt.h     |    2 +-
 include/bsd/nlist.h      |    2 +-
 include/bsd/stdio.h      |    2 +-
 include/bsd/stdlib.h     |    2 +-
 include/bsd/string.h     |    2 +-
 include/bsd/sys/cdefs.h  |    2 +-
 include/bsd/sys/endian.h |    2 +-
 include/bsd/sys/poll.h   |    2 +-
 include/bsd/unistd.h     |    2 +-
 src/fgetln.c             |    2 +-
 src/fpurge.c             |    2 +-
 src/getpeereid.c         |    2 +-
 src/local-elf.h          |    2 +-
 src/setproctitle.c       |    2 +-
 17 files changed, 17 insertions(+), 17 deletions(-)

commit e59ac2c96c4f9299912868cd86fa650639050f78
Author: Guillem Jover <guillem@hadrons.org>
Date:   Fri Mar 23 10:11:57 2012 +0100

    Clarify that the 4-clause BSD licenses are only for man pages

 COPYING |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

commit 7cfa7e4304756151b7987a62fb304300f1205a42
Author: Guillem Jover <guillem@hadrons.org>
Date:   Thu Jan 12 19:13:16 2012 +0100

    Add new man page for tree(3)
    
    Taken from FreeBSD.

 man/Makefile.am |    1 +
 man/tree.3      |  504 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 505 insertions(+)

commit 7620fef70bac36435dae756554f575826cb89f8b
Author: Guillem Jover <guillem@hadrons.org>
Date:   Thu Jan 12 20:56:24 2012 +0100

    Remove UC Berkeley advertising clause
    
    As per <ftp://ftp.cs.berkeley.edu/pub/4bsd/README.Impt.License.Change>.

 man/bitstring.3 |    4 ----
 man/queue.3bsd  |    4 ----
 2 files changed, 8 deletions(-)

commit 08139dd50e6b728a1d9052e852799f60a7f8482d
Author: Guillem Jover <guillem@hadrons.org>
Date:   Thu Jan 12 19:13:16 2012 +0100

    Add new man pages for bitstring(3) and queue(3)
    

libbsd-0.4.2/src/hash/md5hl.c
/**	$MirOS: src/lib/libc/hash/helper.c,v 1.5 2007/05/07 15:21:18 tg Exp $ */
/*	$OpenBSD: helper.c,v 1.8 2005/08/08 08:05:35 espie Exp $	*/

/*
 * ----------------------------------------------------------------------------
 * "THE BEER-WARE LICENSE" (Revision 42):
 * <phk@login.dkuug.dk> wrote this file.  As long as you retain this notice you
 * can do whatever you want with this stuff. If we meet some day, and you think
 * this stuff is worth it, you can buy me a beer in return.   Poul-Henning Kamp
 * ----------------------------------------------------------------------------
 */

#include <sys/param.h>
#include <sys/stat.h>

#include <errno.h>
#include <fcntl.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include <md5.h>

/* ARGSUSED */
char *
MD5End(MD5_CTX *ctx, char *buf)
{
	int i;
	u_int8_t digest[MD5_DIGEST_LENGTH];
#ifdef MD5_DIGEST_UPPERCASE
	static const char hex[] = "0123456789ABCDEF";
#else
	static const char hex[] = "0123456789abcdef";
#endif

	if (buf == NULL && (buf = malloc(MD5_DIGEST_STRING_LENGTH)) == NULL)
		return (NULL);

	MD5Final(digest, ctx);
	for (i = 0; i < MD5_DIGEST_LENGTH; i++) {
		buf[i + i] = hex[digest[i] >> 4];
		buf[i + i + 1] = hex[digest[i] & 0x0f];
	}
	buf[i + i] = '\0';
	memset(digest, 0, sizeof(digest));
	return (buf);
}

char *
MD5FileChunk(const char *filename, char *buf, off_t off, off_t len)
{
	struct stat sb;
	u_char buffer[BUFSIZ];
	MD5_CTX ctx;
	int fd, save_errno;
	ssize_t nr;

	MD5Init(&ctx);

	if ((fd = open(filename, O_RDONLY)) < 0)
		return (NULL);
	if (len == 0) {
		if (fstat(fd, &sb) == -1) {
			close(fd);
			return (NULL);
		}
		len = sb.st_size;
	}
	if ((len < 0) || (off > 0 && lseek(fd, off, SEEK_SET) < 0))
		return (NULL);

	while ((nr = read(fd, buffer,
	    (size_t)(len ? MIN(BUFSIZ, len) : BUFSIZ))) > 0) {
		MD5Update(&ctx, buffer, (size_t)nr);
		if (len > 0 && (len -= nr) == 0)
			break;
	}

	save_errno = errno;
	close(fd);
	errno = save_errno;
	return (nr < 0 ? NULL : MD5End(&ctx, buf));
}

char *
MD5File(const char *filename, char *buf)
{
	return (MD5FileChunk(filename, buf, (off_t)0, (off_t)0));
}

char *
MD5Data(const u_char *data, size_t len, char *buf)
{
	MD5_CTX ctx;

	MD5Init(&ctx);
	MD5Update(&ctx, data, len);
	return (MD5End(&ctx, buf));
}
